### ISSUE TITLE
Invalid conditional references in Encounter data

### ISSUE THREAD

**Issue by alexzautke** [mcuf-idim/fhir-benchmark#1]
Hi @martinhinze,

while importing the Encounter data into our FHIR server, we noticed that that the "individual" elements contain invalid references to Practitioners. It looks like that all encounters contain a "conditional reference" with a lookup on identifier using the us-npi. These references are only allowed in transaction bundles. When importing resources individually they need to be in the correct form of "Practitioner/<id>". Currently our system is rejecting the resources as "?identifier=http://hl7.org/fhir/sid/us-npi|9999922195" is not a valid id and also the "/" after Practitioner is missing.

Could you clarify the intention behind these references or confirm that this is an error in the data?

**Comment by martinhinze** [comment]
Hey @alexzautke,

Thank you for pointing out this important issue. I believe you’re correct, there seems to be an error in the data. To double-check: We currently have key-value pairs like `"reference":"Practitioner?identifier=http://hl7.org/fhir/sid/us-npi|9999947796"`, but these should be `"reference":"Practitioner/5f5d0929-a08d-3239-a309-915f146ea04a"` — is that correct?

Have you encountered any other errors like this? For instance, in the `Encounter` resources, I also see key-value pairs like `"reference":"Location?identifier=https://github.com/synthetichealth/synthea|16162e2e-393c-3162-8e09-a84a60d825c4"`, which would also be an incorrect notation.

This type of conditional reference appears in other NDJSON files as well, and I’m working on how to address this quickly. We tested the benchmark on several servers but didn’t encounter this error before.

**Comment by alexzautke** [comment]
> We currently have key-value pairs like "reference":"Practitioner?identifier=http://hl7.org/fhir/sid/us-npi|9999947796", but these should be "reference":"Practitioner/5f5d0929-a08d-3239-a309-915f146ea04a" — is that correct?

Yes, that's correct.

> Have you encountered any other errors like this? For instance, in the Encounter resources, I also see key-value pairs like "reference":"Location?identifier=https://github.com/synthetichealth/synthea|16162e2e-393c-3162-8e09-a84a60d825c4", which would also be an incorrect notation.

I haven't fully checked everything yet. I just took the first element where I've encountered the error.

**Comment by martinhinze** [comment]
Hello @alexzautke,

I’ve updated the dataset and replaced the conditional references with simple references, as discussed above. The [updated dataset](https://github.com/mcuf-idim/fhir-benchmark-data/blob/replace-conditional-refs/data.tar.gz) is available in a feature branch of the data repository. Could you please check if this works with your server?

You can replace the old dataset with the new one by following these steps:

```bash
# Enter the data directory of the fhir-benchmark on your local machine, e.g.
cd fhir-benchmark/data
# Delete the old data (double-check the working directory before proceeding)
rm ./*.ndjson ./*.ids
# Download the new dataset
curl -L -o data.tar.gz \
https://github.com/mcuf-idim/fhir-benchmark-data/raw/replace-conditional-refs/data.tar.gz &&
tar -xzf data.tar.gz && \
rm data.tar.gz
```

Once that's done, you can run the benchmark as described in the readme. Please let me know if everything works as expected!

**Comment by alexzautke** [comment]
Hi @martinhinze,

in some places the references are still broken. Mainly the .location elements, for example in in Procedure and Immunization.

**Comment by martinhinze** [comment]
I'm sorry to hear that. When you say "broken," do you mean the reference is formatted incorrectly, or is it pointing to a resource that doesn't exist? I ran the following command:
```bash
for file in *.ndjson; do
    echo "Checking $file"
    jq -r '.. | objects | select(has("reference")) | .reference' "$file" | grep -vE '^[A-Za-z]+\/[0-9a-fA-F\-]+$'
done
```
within the data folder and didn't find any matches. Could you provide a few examples of the broken references?

Also, are you unable to run the benchmark at all, or is the server rejecting only some resources?

**Comment by alexzautke** [comment]
Broken as in still contains the wrongly formatted references

**Comment by alexzautke** [comment]
This is resulting in the server running into an exception handling producing errors resulting in a very slow execution of the benchmark.

**Comment by martinhinze** [comment]
Okay, could you provide some examples of the incorrectly formatted references? I need to create a regular expression to catch all of them, and my attempt above didn't yield any results.

**Comment by alexzautke** [comment]
```{"resourceType":"Procedure","id":"48e69573-315a-6834-d5ec-f10e07ccafef","meta":{"profile":["http://hl7.org/fhir/us/core/StructureDefinition/us-core-procedure"]},"status":"completed","code":{"coding":[{"system":"http://snomed.info/sct","code":"1225002","display":"Upper arm X-ray"}],"text":"Upper arm X-ray"},"subject":{"reference":"Patient/b6575849-1be3-ca51-2b18-7554a1dc9544"},"encounter":{"reference":"Encounter/e5c18b7d-63cf-12ae-3af5-d8c1d7192079"},"performedPeriod":{"start":"2014-11-09T01:07:20+00:00","end":"2014-11-09T01:37:20+00:00"},"location":{"reference":"Location?identifier=https://github.com/synthetichealth/synthea|a45ad017-fcd1-3370-a5ce-e6c5143ed315","display":"DIMOCK COMMUNITY HEALTH CENTER, INC"}}```

**Comment by alexzautke** [comment]
That's what I'm seeing in the latest data file

**Comment by alexzautke** [comment]
Or in Immunization:

```
{"resourceType":"Immunization","id":"41620b8d-ea11-a739-f851-a79c2e2793ba","meta":{"profile":["http://hl7.org/fhir/us/core/StructureDefinition/us-core-immunization"]},"status":"completed","vaccineCode":{"coding":[{"system":"http://hl7.org/fhir/sid/cvx","code":"113","display":"Td (adult), 5 Lf tetanus toxoid, preservative free, adsorbed"}],"text":"Td (adult), 5 Lf tetanus toxoid, preservative free, adsorbed"},"patient":{"reference":"Patient/22c8d92f-3598-2608-fb6c-4df57a611df6"},"encounter":{"reference":"Encounter/401f52bb-e120-4a3b-bc40-adb1b057d7fb"},"occurrenceDateTime":"2015-09-01T16:41:29+00:00","primarySource":true,"location":{"reference":"Location?identifier=https://github.com/synthetichealth/synthea|f507dfe8-cf7f-3064-affb-a20f9d13ca34","display":"LIFE LINE COMMUNITY HEALTHCARE"}}
```

**Comment by martinhinze** [comment]
Thank you, I am reviewing the data. The `Procedure` resource above, with the ID `48e69573-315a-6834-d5ec-f10e07ccafef`, comes from the file `Procedure.ndjson`, line 32. In the latest version of the data, the resource appears like this on my end:
```
{"resourceType":"Procedure","id":"48e69573-315a-6834-d5ec-f10e07ccafef","meta":{"profile":["http://hl7.org/fhir/us/core/StructureDefinition/us-core-procedure"]},"status":"completed","code":{"coding":[{"system":"http://snomed.info/sct","code":"1225002","display":"Upper arm X-ray"}],"text":"Upper arm X-ray"},"subject":{"reference":"Patient/b6575849-1be3-ca51-2b18-7554a1dc9544"},"encounter":{"reference":"Encounter/e5c18b7d-63cf-12ae-3af5-d8c1d7192079"},"performedPeriod":{"start":"2014-11-09T01:07:20+00:00","end":"2014-11-09T01:37:20+00:00"},"location":{"reference":"Location/a45ad017-fcd1-3370-a5ce-e6c5143ed315","display":"DIMOCK COMMUNITY HEALTH CENTER, INC"}}
```
So, in the new version of the code, we have "reference":"Location/a45ad017-fcd1-3370-a5ce-e6c5143ed315", whereas in the old version, we had "reference":"Location?identifier=https://github.com/synthetichealth/synthea|a45ad017-fcd1-3370-a5ce-e6c5143ed315".

As mentioned earlier, the latest version of the data is still stored in a feature branch:

https://github.com/mcuf-idim/fhir-benchmark-data/raw/replace-conditional-refs/data.tar.gz

Are we looking at the same dataset?

**Comment by alexzautke** [comment]
Oh sorry, I thought you'd pushed that to the main!

**Comment by alexzautke** [comment]
Let me try again

**Comment by martinhinze** [comment]
Yes, please let me know if everything works as expected. I wanted to wait for confirmation from your side before merging into the main branch. 😅

**Comment by alexzautke** [comment]
Seems to work now!

**Comment by martinhinze** [comment]
I’m glad to hear that! Thank you for pointing out the issue.
